subsys/bluetooth/host/smp.c

```diff
@@ -3954,12 +4017,20 @@ static uint8_t smp_security_request(struct bt_smp *smp, struct net_buf *buf)
 
 static uint8_t generate_dhkey(struct bt_smp *smp)
 {
+	int err;
+
 	if (IS_ENABLED(CONFIG_BT_SMP_OOB_LEGACY_PAIR_ONLY)) {
 		return BT_SMP_ERR_UNSPECIFIED;
 	}
 
-	if (bt_dh_key_gen(smp->pkey, bt_smp_dhkey_ready)) {
-		return BT_SMP_ERR_UNSPECIFIED;
+	err = bt_dh_key_gen(smp->pkey, bt_smp_dhkey_ready);
+	if (err) {
+		if (err != -EALREADY) {
+			BT_ERR("Failed to generate DHKey");
+			return BT_SMP_ERR_UNSPECIFIED;
+		}
```

Comment:
Use `atomic_test_and_set_bit` ahead of this key_gen call instead? 
This is nearly identical to `smp_dhkey_next`, and has enough behavior to hold potential bugs. Perhaps they should share their implementation?